### PR TITLE
Fix /jamute muting too many groups

### PR DESCRIPTION
**Pull request by awoo-civ**
Fixes https://github.com/DevotedMC/JukeAlert/issues/31

A workaround, might hurt performance, might not. A proper solution would involve changing the way the /jamute group data is stored in the DB.

### CODE REVIEW

**Comment by ProgrammerDan** [comment]
I'm not entirely sure I grok this yet, I'll have to look at the db a bit

**Comment by awoo-civ** [comment]
Muted groups are stored in the DB like this:

uuid | space-separated-groups
-|-
kjfdjf8dsjf8ds | test test123 testtest
8dksjc8dskfd | somegroup hi test

The code checks which uuids ignore a particular group by checking if the group's name is anywhere in the space-separated-groups string:
 
- https://github.com/DevotedMC/JukeAlert/blob/master/src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java#L1647

```java
String sql = "%" + ignoredGroup + "%";
getIgnoreUUIDSStmt.setString(1, sql);
ResultSet set = getIgnoreUUIDSStmt.executeQuery();
```
- https://github.com/DevotedMC/JukeAlert/blob/master/src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java#L593

```java
getIgnoreUUIDSStmt = db.prepareStatement(String.format(
    "SELECT * FROM %s WHERE muted_groups LIKE ? ;", mutedGroupsTbl));
```

So the root of the problem is that muted groups are stored as a string instead of in a way that the DB understands as a set of values.

My fix checks if the muted group actually exists in each `uuid: muted_groups` key-value pair received from the DB.

**Comment by ProgrammerDan** [comment]
Oh wow, I'd forgotten how terrible that is. Ok, thanks for the explain!

File: src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java [hunk 1]
```diff
@@ -1632,11 +1632,19 @@ public boolean updateMutedGroups(UUID uuid, String group2Mute) {
 			return null;
 		} else {
 			// Add the first value
-			ignoringUsers.add(set.getString(1));
+			String[] mutedGroups = set.getString(2).split("\\s+");
+			Set<String> mutedGroupsSet = new HashSet<String>(Arrays.asList(mutedGroups));
```
Comment (ProgrammerDan) [inline src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java]: Save a Set, use List instead.

Arrays.asList(mutedGroups) will return an object of interface type List<String> -- this also has "contains()".

As written right now, you're taking a lightweight array, iterating over it to make a List, then iterating over it again to make a set, so you can get log(n) performance on the contains().

Slight improvement to just use the Arrays.asList() result and use Contains on it, and uses less memory.

**Review by ProgrammerDan — CHANGES_REQUESTED** [review]
Thought I published this hours ago, sorrry!

File: src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java (hunk 1 above)
Comment (awoo-civ) [reply src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java]: Ye I just wanted the O(1) on .contains(), probably irrelevant as there's only so many groups a player will have muted.

BTW, Arrays.asList() is O(1) and so is the .contains() on HashSet no?

File: src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java [hunk 2]
```diff
@@ -1632,11 +1632,19 @@ public boolean updateMutedGroups(UUID uuid, String group2Mute) {
 			return null;
 		} else {
 			// Add the first value
-			ignoringUsers.add(set.getString(1));
+			String[] mutedGroups = set.getString(2).split("\\s+");
+			List<String> mutedGroupsSet = new ArrayList<String>(Arrays.asList(mutedGroups));
```
Comment (ProgrammerDan) [inline src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java]: I probably wasn't clear.

Arrays.asList(mutedGroups) --- all by itself, don't wrap it with anything else.

File: src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java (hunk 1 above)
Comment (ProgrammerDan) [reply src/main/java/com/untamedears/JukeAlert/storage/JukeAlertLogger.java]: Arrays.asList() appears to be O(1) yea.  new Set( _list_ ) is O(n) -- so at least in that original version you had O(1) + O(n) to get O(1) retrieval (best case). With streamlined, you just have O(1) to get O(n) retrieval, plus no hash computations during set setup, etc. ... largely pedantry at this scale, but always good to keep goal in mind. Spending O(n) for O(1) retrieval makes sense if you do it more then once. In this case where you're setting up a set just to throw it away after a single .contains() -- wouldn't make sense to do the hash setup.

**Pull request merged**